### ISSUE TITLE
Potential reasons for `NaN` during training

### ISSUE THREAD

**Issue by michaeldeistler** [jaxleyverse/jaxley#317]
Below are several reasons for experiencing `NaN`, ranked from most to least likely. If you have tried out all of the options below, we would be happy to receive a bug report (as an [issue on Github](https://github.com/jaxleyverse/jaxley/issues/new)) with the following information:
- are you using your own channel and/or synapse models? If yes, ideally share your channel and synapse models.
- do the `NaN` occur during training or during simulation?
- are you using point neurons or neurons with morphology? If the neurons have a morphology, did you define the morphology from scratch or did you read it from an SWC file?

### 1. You are using `float32`.

In most cases where I have encountered `NaN` so far, `NaN` were resolved by switching to `float64`:
```python
from jax import config
config.update("jax_enable_x64", True)
```

### 2. The mechanisms (channels or synapses) are unstable.

The most likely reason for this is that the channels you are using contain a `jnp.exp()` whose input gets very large (>100.0) such that the result will be `inf`. For example, our initial implementation of channels had given `NaN` when a strong negative stimulus was inserted, such that the neuron was very strongly hyperpolarized (voltages below -200mV). You can prevent this by chaning `jnp.exp()` to `save_exp` from `jaxley.solver_gate.py`.

### 3. The `ParamTransform` saturates:
```python
transform = jx.ParamTransform(lowers={"x": -1.0}, uppers={"x": 1.0})
def tf(params):
    return jnp.sum(transform.forward(params)[0]["x"])

# Interestingly, only negative values return `NaN` gradient, positive values return `0`.
p = [{"x": jnp.asarray([-100.0])}]

tf_grad_fn = value_and_grad(tf)
print(tf(p))
print(tf_grad_fn(p))
```
To debug this, print the maximum value of the transformed params after every gradient update:
```python
parameters = net.get_parameters()
opt_params = transform.inverse(parameters)

leaves, _ = tree_util.tree_flatten(x)
max_val = jnp.max([jnp.abs(leaf) for leaf in leaves])
```
If `max_val > 50.0` then you are probably in trouble.
